Added WatcherPotionEffect and LastHealth to 1.12.x

XZcutable opened the pull request:
WatcherPotionEffect is a watcher added to make sure that the client and
server are always in agreement with which potion effects are active. The
client does **not** influence what potion effects the server has on any given
LivingEntity.

LastHealth acts similarly to Heath as it is always one step behind what
health was. This data is persistent and written to the nbt and a getter and setter were added.

I have created mod that uses both of these features called [Damage Indicator](https://github.com/Z-Doctor/DamageIndicator) that uses these values to draw the entities health above its head as it would look for the player, including potion effects.

LastHealth is needed so that a ghostly heart can be drawn where the health was at before the previous injury.

I have tested as well as I am able and have not found any problems.

Here is a [forum post](http://www.minecraftforge.net/forum/topic/66408-suggestion-for-a-few-small-changes/?do=findComment&comment=318367) as to why I felt that these additions were necessary as well as some pictures of the mod. **Note:** I have changed some of the code given in that post to better integrate it into the forge code and required fewer changes than I originally thought.

XZcutable committed:
Added WatcherPotionEffect and LastHealth

WatcherPotionEffect is a watcher added to make sure that the client and
server are always in agreement with which potion effects are active. The
client does not influence what potion effects the server has on any
given
LivingEntity.

LastHealth acts similarly to Heath as it is always one step behind what
health was. This data is persistent and written to the nbt and a getter
and setter were added.

I have created mod that uses both of these features called Damage
Indicator that uses these values to draw the entities health above its
head as it would look for the player, including potion effects.

LastHealth is needed so that a ghostly heart can be drawn where the
health was at before the previous injury.

I have tested as well as I am able and have not found any problems.

Here is a forum post as to why I felt that these additions were
necessary as well as some pictures of the mod. Note: I have changed some
of the code given in that post to better integrate it into the forge
code and required fewer changes than I originally thought.

covers1624 commented:
What happens if a vanilla client connects to the server with this patch? Why can't you do this yourself with events and your own packets? Also Please go over your formatting, i see some incorrect indenting and random newlines.

XZcutable commented:
I'll look over the formatting again, I was sure I followed the format rules. I'm not sure about random new lines, but I'll look at it again. If a vanilla client connects to a server with this patch then it wouldn't do much. As the code is, there is only 1 part that actually does anything on the client side and that is just read an array and unpack it. The server would only set this variable.

If the client had this version and the server didn't, then simply it would work as it does now. The potion effects would never be updated. If a client didn't have this version, it wouldn't know to look for these changes and would act the same.

As to why I can't do this myself. I can. I have successfully created an entity that would update its potion effects correctly, but that does nothing for anyone for every other entity that would still have this problem. There are no events to attach to when a potion effect is added or removed. The only relevant event I found in the code was PotionColorCalculationEvent, but that wouldn't work as that is only called when a potion is added, not removed and not called at all if there are no active effects.

XZcutable commented:
Here is a break down of everything done related to WATCHER_POTION_EFFECTS:
**Removed:** Nothing
**Added:**
- A DataParameter<NBTTagCompound> called WATCHER_POTION_EFFECTS
- A line in entityInit that registers WATCHER_POTION_EFFECTS with a new NBTTagCompound
- 11 lines in updatePotionMetadata that do the following:
1. Checks to make sure the world is the server and does nothing otherwise
2.  Creates a local variable nbttaglist with a new NBTTagList
3. Goes through each PotionEffect given from the local varaible collection (which was already there) and writes its affect to a NBTCompound and adds that to the nbttaglist
4. Creates a new local NBTTagCompound named compound, calls setTag setting the nbttag list as "ActiveEffects" and sets this NBTCompund to WATCHER_POTION_EFFECTS.
- A line is added to resetPotionEffectMetadata that sets WATCHER_POTION_EFFECTS to a new NBTTagCompund
- 14 Lines are added to notifyDataManagerChange after the check with HAND_STATES that does the following:
1. Checks if the key equals WATCHER_POTION_EFFECTS and if the world is the client and does nothing otherwise
2. Creates a local variable nbttaglist a NBTTagList from the WATCHER_POTION_EFFECTS compound and the list under "ActiveEffects"
3. Reads each NBTTagCompund from the NBTTagList and creates a PotionEffect from the NBTTagCompund
4. It checks to make sure that the PotionEffect is not null and passes that potion effect to addPotionEffect.

kashike commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -1,22 +1,38 @@
+     private static final DataParameter<Boolean> field_184634_g = EntityDataManager.<Boolean>func_187226_a(EntityLivingBase.class, DataSerializers.field_187198_h);
+     private static final DataParameter<Integer> field_184635_h = EntityDataManager.<Integer>func_187226_a(EntityLivingBase.class, DataSerializers.field_187192_b);
++    private static final DataParameter<Float> LAST_HEALTH = EntityDataManager.func_187226_a(EntityLivingBase.class, DataSerializers.field_187193_c);
++    private static final DataParameter<NBTTagCompound> WATCHER_POTION_EFFECTS = EntityDataManager.func_187226_a(EntityLivingBase.class, DataSerializers.field_192734_n);
```
It'd be great if you could make this an `Optional<NBTTagCompound>` instead, and save creating a compound tag for *every single living entity that is constructed*.

XZcutable replied:
I am not familiar with the Optional class, but I assume you are talking about the com.google.common.base.Optional class? So I don't quite understand your suggestion or how I would go about implementing it. I'll try playing around with it for a bit to see if I can grasp it.

XZcutable replied:
I also use DataParameter because I rely on its ability to automatically send and receive the relevant packets to keep the value consistent across the server and client. Can Optional do that?

LexManos replied:
DataParamaeters are not allows as they break network compatibility with vanilla and older clients.

XZcutable replied:
@LexManos 
What does that mean? Are you saying my patch is invalid since it adds 2 new data parameters?

XZcutable commented:
Here is a break down of everything done related to LAST_HEALTH:
**Removed:** Nothing
**Added:**
- A DataParameter<Float> called LAST_HEALTH
- A line in entityInit that registers LAST_HEALTH with a the same value as health: 1.0F
- A line in setHealth that calls setLastHealth and passes the current health before the heath is changed
- A method called getLastHealth that returns a float from LAST_HEALTH
- A method called setLastHealth that sets LAST_HEALTH with a given float
- One line in writeEntityToNBT after it writes the health to the NBT it writes a float with the tag "LastHealth" to the nbt from the value retrieved from getLastHealth 
- 3 lines in readEntityFromNBT that is similar to how health is read. It checks if the tag has the key "LastHealth" and then calls setLastHealth with that value

Z-Doctor commented:
@covers1624 
I'll look over the formatting again, I was sure I followed the format rules. I'm not sure about random new lines, but I'll look at it again. If a vanilla client connects to a server with this patch then it wouldn't do much. As the code is, there is only 1 part that actually does anything on the client side and that is just read an array and unpack it. The server would only set this variable.

If the client had this version and the server didn't, then simply it would work as it does now. The potion effects would never be updated. If a client didn't have this version, it wouldn't know to look for these changes and would act the same.

As to why I can't do this myself. I can. I have successfully created an entity that would update its potion effects correctly, but that does nothing for anyone for every other entity that would still have this problem. There are no events to attach to when a potion effect is added or removed. The only relevant event I found in the code was PotionColorCalculationEvent, but that wouldn't work as that is only called when a potion is added, not removed and not called at all if there are no active effects.

XZcutable committed:
Moved things closer together

Moved things closer together removing 16 unnecessary lines of code from
the patch

XZcutable commented:
After reading the [wiki](https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge), I've realized that I should attempt to keep the number of new lines as small as possible and forgo the formatting in existing classes (such as EntityLivingBases). This is the result of that along with moving most new entries to the end of the method (such as LAST_HEALTH in readNBT and writeNBT).

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -47,15 +63,40 @@
+             this.func_70606_j(p_70037_1_.func_74760_g("Health"));
+         }
+-
++        
```
wat?

XZcutable replied:
I'm unsure what you are asking, I am simply trying my best to make the follow the given specifications and use genPatches to generate the patches.

covers1624 replied:
You have a change here you need to get rid of, its just adding indentation.

covers1624 commented:
Keep things small, yes, but readable. Don't change existing formatting, don't remove blocks of code inside if's and such, new methods and large blobs of code should be properly formatted, not squished up into single lines, that makes it even more unmaintainable. The idea is to make the patches as simple as possible to make updating things easy, so who ever is updating it doenst have to sift through formatting changes.

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -66,9 +107,23 @@
 +            this.field_70180_af.func_187227_b(field_184634_g, event.areParticlesHidden());
 +            this.field_70180_af.func_187227_b(field_184633_f, event.getColor());
              this.func_82142_c(this.func_70644_a(MobEffects.field_76441_p));
++            if(!field_70170_p.field_72995_K) { NBTTagList nbttaglist = new NBTTagList();
```
Oh god That's nasty

XZcutable replied:
I am sorry, I am new to this so I am having trouble seeing the problem.

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -77,15 +132,29 @@
+         return ((Float)this.field_70180_af.func_187225_a(field_184632_c)).floatValue();
+     }
++    
+ 
```
again?

XZcutable replied:
Another line, I think?

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -77,15 +132,29 @@
+         this.field_70180_af.func_187227_b(field_184632_c, Float.valueOf(MathHelper.func_76131_a(p_70606_1_, 0.0F, this.func_110138_aP())));
+     }
++    
++    public float getLastHealth() { return ((Float)this.field_70180_af.func_187225_a(LAST_HEALTH)).floatValue(); }
```
Do you even look at this before you commit?

XZcutable replied:
I look at the java code, am I supposed to look at the patch?

LexManos commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -395,7 +464,21 @@
++            NBTTagList nbttaglist = this.field_70180_af.func_187225_a(WATCHER_POTION_EFFECTS).func_150295_c("ActiveEffects", 10);
++    		field_70713_bf.clear();
++    		for (int i = 0; i < nbttaglist.func_74745_c(); ++i) {
++    			NBTTagCompound nbttagcompound = nbttaglist.func_150305_b(i);
```
...

XZcutable replied:
Now I am really stumped about this one. This is definitely needed. Should I have put it in one line?

XZcutable commented:
When I said forgo formatting I meant on the new code that I added trying to keep it readable and not taking up to much space. I didn't touch the format of the other code and made sure not to auto format the code with eclipse.

covers1624 commented:
`new methods and large blobs of code should be properly formatted`, I suggest you check out some other patch files and additions forge adds. All of lex's comments are hitting on this note.

LexManos commented:
Overall your formatting is horrid, you add extra blank lines to the patch for no reason, your formatting for basic ifs and other code blocks is nothing like the surrounding code.
Your entire approach of encoding everything to a NBT and checking the entire NBT with each other is horrid/inefficient.
The actual active potion effects are synced to the client. Thats how the client knows to rendering the nausia/blindness effects, as well as the particles for other potions.

As for the 'last health' concept, this is a very niche value. You can easily achieve this yourself by adding a cap that holds this value to entities and updating it in your render handler.

XZcutable commented:
I'm sorry that my formatting was wrong. I'll try to fix it. Perhaps I wasn't clear, the issue is not that the client is unaware of the player's potion effects, but every other living entity that is not the player. As for the lastHealth value. Perhaps you are right that it is a niche value. Not as many modders would need this in the beginning, but what about the ones who do? However, I will focus on adding just one aspect for now.

LexManos commented:
Alright, looking more into it you are right, but only in a sense. Mojang 1/2 implemented sending all living entity potion packets. This doesn't excuse the rest of the issues with this PR. And your general concept is just wrong when it comes to how to fix this.
Vanilla already has the processes to sync the potion information, it would just need to be synced to the client using the mechanics the player uses.
Tho, I suspect Mojang DIDNT do this to prevent massive network spam for what will most likely be useless data to the client. So there would need to be some guard to make sure the client cares about this.
